client/order/orderbook.go

```diff
@@ -0,0 +1,226 @@
+// the order book mutex held.
+func (ob *OrderBook) removeOrder(order *msgjson.BookOrderNote) error {
+	switch order.Side {
+	case msgjson.BuyOrderNum:
+		ordersAtPrice, ok := ob.buys[order.Rate]
+		if !ok {
+			return fmt.Errorf("no buy orders found at price %v", order.Rate)
+		}
+
+		for idx := len(ordersAtPrice) - 1; idx >= 0; idx-- {
+			if bytes.Equal(ordersAtPrice[idx].OrderID, order.OrderID) {
+				ordersAtPrice = append(ordersAtPrice[:idx], ordersAtPrice[idx+1:]...)
+				ob.buys[order.Rate] = ordersAtPrice
+				break
+			}
+		}
+
+	case msgjson.SellOrderNum:
+		ordersAtPrice, ok := ob.sells[order.Rate]
+		if !ok {
+			return fmt.Errorf("no sell orders found at price %v", order.Rate)
+		}
+
+		for idx := len(ordersAtPrice) - 1; idx >= 0; idx-- {
+			if bytes.Equal(ordersAtPrice[idx].OrderID, order.OrderID) {
+				ordersAtPrice = append(ordersAtPrice[:idx], ordersAtPrice[idx+1:]...)
+				ob.sells[order.Rate] = ordersAtPrice
+				break
+			}
+		}
+	}
+
+	delete(ob.orders, order.OrderID.String())
+
+	return nil
+}
+
+// Initialize synchronizes a client tracked order book with an order
+// book snapshot.
+func (ob *OrderBook) Initialize(snapshot *msgjson.OrderBook) error {
+	ob.mtx.Lock()
+	defer ob.mtx.Unlock()
+
+	if ob.synced {
+		return fmt.Errorf("reinitializing a synced order book")
+	}
+
+	ob.orders = make(map[string]*msgjson.BookOrderNote, len(snapshot.Orders))
+	ob.buys = make(map[uint64][]*msgjson.BookOrderNote)
+	ob.sells = make(map[uint64][]*msgjson.BookOrderNote)
+
+	for _, order := range snapshot.Orders {
+		ob.addOrder(order)
+	}
+
+	ob.seq = snapshot.Seq
+	ob.marketID = snapshot.MarketID
+	ob.synced = true
+
+	return nil
+}
+
+// Book adds a new order to the order book.
+func (ob *OrderBook) Book(order *msgjson.BookOrderNote) error {
+	ob.mtx.Lock()
+	defer ob.mtx.Unlock()
+
+	if !ob.synced {
+		return fmt.Errorf("order book is unsynced")
+	}
+
+	if ob.marketID != order.MarketID {
+		return fmt.Errorf("market id mismatch, expected %s got %s",
+			ob.marketID, order.MarketID)
+	}
+
+	expectedSeq := ob.seq + 1
+	if expectedSeq != order.Seq {
+		ob.synced = false
+		return fmt.Errorf("order book out of sync, expected %d got %d",
+			expectedSeq, order.Seq)
+	}
+
+	ob.seq++
+	ob.addOrder(order)
+	return nil
+}
+
+// Unbook removes an order from the order book.
+func (ob *OrderBook) Unbook(unbookOrder *msgjson.UnbookOrderNote) error {
+	ob.mtx.Lock()
+	defer ob.mtx.Unlock()
+
+	if !ob.synced {
+		return fmt.Errorf("order book is unsynced")
+	}
+
+	if ob.marketID != unbookOrder.MarketID {
+		return fmt.Errorf("market id mismatch, expected %s got %s",
+			ob.marketID, unbookOrder.MarketID)
+	}
+
+	order, ok := ob.orders[unbookOrder.OrderID.String()]
+	if !ok {
+		return fmt.Errorf("no order found with id %v",
+			unbookOrder.OrderID.String())
+	}
+
+	expectedSeq := ob.seq + 1
+	if expectedSeq != order.Seq {
+		ob.synced = false
+		return fmt.Errorf("order book out of sync, expected %v got %v",
+			expectedSeq, order.Seq)
+	}
```

Comment:
Yep. Should be checking `unbookOrder.Seq`.

Reply:
This is even more tricky actually. Because of the syncing requirements on the server, it's possible to receive your first notification before the initial book comes, so there needs to be a small cache.

Reply:
That'd mean it'd be possible to get a book notification before the initial books comes in? That seems off to me. The notifications received act on the book so wouldn't every notification received before the book be invalid since there is no book to give it context?

Reply:
The large order book can take time to encode. Since the client cannot miss any notifications, the client is added to the subscription map just before the order book is sent. The order book is sent asynchronously. That's means that it's entirely possible for a notification to be sent while the order book is still being encoded. That notification would have a sequence ID after the book still being received. 

It's not uncommon for the client to be required to handle out of order updates. See https://github.com/decred/dcrdata/blob/ddb4fb4a56cc7333fa4377809160e4e4ecfb4b08/exchanges/exchanges.go#L1470 for an example of how it's handled for syncing a Bittrex order book.

Reply:
https://github.com/decred/dcrdex/blob/fc0c0d42a745e18c59df22553451fc886e63f2f5/server/market/bookrouter.go#L294-L295